### ISSUE TITLE
KeytoolCertificateGenerator: Improve API and implementation

### ISSUE THREAD

**Issue by k-wall** [kroxylicious/kroxylicious-junit5-extension#131]
`KeytoolCertificateGenerator` was a great start that has allowed us to quickly get testing TLS use-cases.  However, its API is a [bit weird](https://github.com/kroxylicious/kroxylicious-junit5-extension/blob/5c5501d5d4bd5bdd566be2e3d4d537ffaa8e218f/impl/src/test/java/io/kroxylicious/testing/kafka/common/KeytoolCertificateGeneratorTest.java#L50)  and it relies on shelling out to command line tools.

We should think about:

* Do we value the integration tests having the ability to generate TLS certificates on the fly?  Maybe static key material would be good enough.
* Shelling out to command line tools makes the test suite slower, introduces fragility and erects barriers to participation.  Developers need to keep their environments in sync with the same version of the command line tools.  This isn't really reasonable for open source development.
* Self-signed certificate generation in Java isn't possible using the public APIs.  We could adopt Bouncy Castle as a test dependency.  We could then take an approach similar to Strimzi https://github.com/strimzi/strimzi-kafka-operator/blob/main/systemtest/src/main/java/io/strimzi/systemtest/security/SystemTestCertAndKeyBuilder.java

Also - as originally [discussed on #120](https://github.com/kroxylicious/kroxylicious-junit5-extension/issues/120), it would be desirable if the generator supported other keystore formats, and the generated key material was available over an API to support the testing of richer TLS use-cases.  It would also be good if it supported things like generating expired certificates, and certificate chains.

**Comment by franvila** [comment]
I totally agree. This wasn't the best approach, but it was the only one I found to generate the certs. I didn't know bouncy castle but we could follow the same approach than strimzi if it fits to our goals.

**Comment by k-wall** [comment]
> I totally agree. This wasn't the best approach, but it was the only one I found to generate the certs. I didn't know bouncy castle but we could follow the same approach than strimzi if it fits to our goals.

Like i say, this isn't meant as a criticism of existing work.  It is just a realisation that we need something better.

**Comment by robobario** [comment]
Just bumping this as I'm looking at Vault TLS testing, Vault wan's a private key in PEM form which looks like we'd have to use a separate tool to extract if we start with Keystore as the source of keys. 

For this case we might want to generate the private key and self-signed-cert in PEM with bouncycastle and import the cert into a java store.

**Comment by k-wall** [comment]
We might want to look at https://netty.io/4.2/api/io/netty/pkitesting/CertificateBuilder.html

**Comment by tombentley** [comment]
This was not fixed by #544. #544 added the `KeystoreManager`, tests for it, and deprecated `KeytoolCertificateGenerator`. It didn't fix any use sites.

**Comment by robobario** [comment]
From #649, we are trying to figure out the path forward for this generator.

> It seems to me that it would be better to have a one piece of test support infrastructure for generating certs and keystores, rather than two. That would need:

I think this is the first question to figure out, do we want this to be public API, currently all these classes are in the `testing-impl` module, which is theoretically an implementation detail "Provides two implementations of KafkaCluster and some common constraint annotations for clusters.".

This is probably the first thing we need to figure out. Do we want the junit extension to:
1. offer a general purpose keystore builder as a facility in it's own right
2. keystore construction is an internal detail of this extension used to provision Kafka/clients.

If we go for 1. have it in here then as we find extensions, we might need in `kroxylicious/kroxylicious` to do a juggling to maybe implement them in the proxy until they are available in a released version of the extension.

Maybe it would be better to absorb this tiny amount of duplication and have the junit extension specialized on the details of provisioning Kafka and clients with TLS/mTLS support, with kroxylicious/kroxylicious not depending on this class.

Though I can see that you could say that TLS keystores is something that is part of Kafka testing and thus justified living in the extension, and if we find novel requirements in the proxy for Kafka client and server TLS configuration then it would be worth sharing with the community. (Tom has noted we don't have any open source dependants in mvn central, but I think we must have some proprietary users, we just don't know).

I'd lean towards it being made more firmly an extension implementation detail and live with a little duplication. If we ever get requests for more TLS customization from the community we could move in that direction.
